### PR TITLE
HPCC-22869 Lightweight join must always ensure sides are ungrouped

### PR DESCRIPTION
**Pull request by jakesmith**
Signed-off-by: Jake Smith <jake.smith@lexisnexisrisk.com>

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Comment by jakesmith** [comment]
@ghalliday - pushed to add a regression test.
Please review

**Comment by jakesmith** [comment]
@ghalliday @richardkchapman - should be this be 7.2.x?

**Comment by richardkchapman** [comment]
Has any 7.2.x user reported it?

**Comment by jakesmith** [comment]
yes, the only report so far was on 7.2.36

**Comment by richardkchapman** [comment]
Bit nervous about adding bugfixes that change the result in a point release. Is there a workaround?

**Comment by jakesmith** [comment]
They succeeded in workaround around by adding a DEGROUP on the RHS of the join - not sure if guaranteed to remain once optimized though.

File: thorlcr/activities/join/thjoinslave.cpp [hunk 1]
```diff
@@ -444,10 +444,7 @@ class JoinSlaveActivity : public CSlaveActivity, implements ILookAheadStopNotify
             leftInput->getMetaInfo(info);
             if (info.totalRowsMax==0) 
                 isemptylhs = true;
-            if (rightpartition)
-                leftStream.set(leftInputStream); // already ungrouped
-            else
-                leftStream.setown(createUngroupStream(leftInputStream));
+            leftStream.setown(createUngroupStream(leftInputStream));
```
Comment (ghalliday) [inline thorlcr/activities/join/thjoinslave.cpp]: Is there any way of only doing this if the stream is grouped?

File: testing/regress/ecl/lightweightjoin.ecl [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+############################################################################## */
+
+// Lightweight join is only implemented in Thor
+//noroxie
+//nohthor
```
Comment (ghalliday) [inline testing/regress/ecl/lightweightjoin.ecl]: what happens if it is run in roxie/hthor?

**Review by ghalliday — COMMENTED** [review]
@jakesmith a couple of questions

File: thorlcr/activities/join/thjoinslave.cpp (hunk 1 above)
Comment (jakesmith) [reply thorlcr/activities/join/thjoinslave.cpp]: I guess it is an extra virtual call per row,
and I think I can test input->isGrouped() to make it conditional

File: testing/regress/ecl/lightweightjoin.ecl (hunk 2 above)
Comment (jakesmith) [reply testing/regress/ecl/lightweightjoin.ecl]: it will run and match, but appears not generate a lightweight join..

**Comment by jakesmith** [comment]
@ghalliday - have changed it so conditional on isGrouped()

Not sure whether the new regression query should bother to run on hthor or roxie ?

**Comment by ghalliday** [comment]
The change looks good to me.  I probably would enable the tests for roxie/hthor, even though they don't generate that specific activity kind.
Need to decide which version to merge this in to.

**Comment by jakesmith** [comment]
@ghalliday - have change the regression test so also runs on hthor and roxie (and squashed that change in).

**Pull request merged**